### ISSUE TITLE
Properly tag exported symbols and compile with -fvisibility=hidden

### ISSUE THREAD

**Issue by onitake** [Ultimaker/libArcus#80]
We've been having lots of build issue with leaked internal symbols and dpkg's built-in symbol list verification.
Here's a recent bug report, for example: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=914953
And some general info on how the check works and why it's important: https://wiki.ubuntu.com/stefanlsd/dpkg-gensymbols

Basically, libArcus.so exports all its symbols, internal and external, which means that they leak into code linked against it and cause lots of symbol table changes every time the library is rebuilt - even if the API didn't change at all.

Exported internal symbols are less of a problem with C code, but can be a big bother for C++, as template code may easily lead to many different symbols between compiler versions, dependencies, CPU architectures, etc.

I propose that Ultimaker adopts tagged symbols for all its C++ libraries and, by default, no symbols are exported. This change will also make it much easier track and detect incompatible API changes between library versions - which is the whole point of the dpkg-gensymbols check.
This can be easily achieved in a portable manner with cmake: https://stackoverflow.com/questions/17080869/what-is-the-cmake-equivalent-to-gcc-fvisibility-hidden-when-controlling-the-e/31157258#31157258
TL;DR, add `set_target_properties(Arcus PROPERTIES CXX_VISIBILITY_PRESET hidden)`.
There also seems to be a cmake function that generates export/import macros: `generate_export_header(Arcus)` - that may or may not be useful.

There's some information on how to accomplish explicit symbol exports on the GCC wiki:
http://gcc.gnu.org/wiki/Visibility
And: https://docs.microsoft.com/en-us/previous-versions/3y1sfaz2(v=vs.140)

There's also the C++11 attribute syntax, but it's only supported in gcc and clang (`[[gnu::dllexport]]`): https://clang.llvm.org/docs/AttributeReference.html#dllexport
For VC++, the declspec syntax must be used: https://docs.microsoft.com/en-us/cpp/cpp/attributes?view=vs-2017

**Comment by Ghostkeeper** [comment]
So `generate_export_header(Arcus)` would create a macro `Arcus_EXPORT` so that we can mark every public class with

    extern "C" Arcus_EXPORT SomeClass { ... }

Did I get that right? I don't have any experience with this sort of thing.

**Comment by LipuFei** [comment]
For dev, see JIRA ticket: CURA-6009

**Comment by onitake** [comment]
@Ghostkeeper You'd still have to include the generated Arcus_export.h and make sure it's installed together with the headers. The file name is customisable.

I don't think the extern "C" makes any sense for classes though - the cmake docs show the macros should be used: https://cmake.org/cmake/help/v3.0/module/GenerateExportHeader.html

If there are any methods in the class that shouldn't be exported, you should also mark those with ARCUS_NO_EXPORT.

**Comment by onitake** [comment]
So, it turns out an export macro was already in place, in ArcusExport.h, to make DLL building work.

I propose the following patch to make it useful for other operating systems then:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 072ec3c..578cdd4 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -3,6 +3,7 @@ cmake_minimum_required(VERSION 2.8.12)
 
 include(GNUInstallDirs)
 include(CMakePackageConfigHelpers)
+include(GenerateExportHeader)
 
 option(BUILD_PYTHON "Build " ON)
 option(BUILD_EXAMPLES "Build the example programs" ON)
@@ -118,6 +119,12 @@ set_target_properties(Arcus PROPERTIES
     SOVERSION ${ARCUS_SOVERSION}
     PUBLIC_HEADER "${arcus_HDRS}"
     DEFINE_SYMBOL MAKE_ARCUS_LIB
+    CXX_VISIBILITY_PRESET hidden
+    VISIBILITY_INLINES_HIDDEN 1
+)
+
+generate_export_header(Arcus
+    EXPORT_FILE_NAME src/ArcusExport.h
 )
 
 if(BUILD_EXAMPLES)

diff --git a/src/ArcusExport.h b/src/ArcusExport.h
deleted file mode 100644
index 8073344..0000000
--- a/src/ArcusExport.h
+++ /dev/null
@@ -1,14 +0,0 @@
-#ifndef ARCUS_EXPORT_H
-#define ARCUS_EXPORT_H
-
-#if _WIN32
-    #ifdef MAKE_ARCUS_LIB
-        #define ARCUS_EXPORT __declspec(dllexport)
-    #else
-        #define ARCUS_EXPORT
-    #endif
-#else
-    #define ARCUS_EXPORT
-#endif
-
-#endif

diff --git a/.gitignore b/.gitignore
index 247e820..e2bde24 100644
--- a/.gitignore
+++ b/.gitignore
@@ -66,3 +66,6 @@ nbproject/*
 # SIP generated code
 /python/sipArcuspart*.cpp
 /python/sipAPIArcus.h
+
+# Generated headers
+/src/ArcusExport.h
```
After applying this, you should review each header file and make sure that the set of exported symbols is correct.

**Comment by Ghostkeeper** [comment]
After applying that change, I get the following file as ArcusExport.h:

```
#ifndef ARCUS_EXPORT_H
#define ARCUS_EXPORT_H

#ifdef ARCUS_STATIC_DEFINE
#  define ARCUS_EXPORT
#  define ARCUS_NO_EXPORT
#else
#  ifndef ARCUS_EXPORT
#    ifdef MAKE_ARCUS_LIB
        /* We are building this library */
#      define ARCUS_EXPORT __attribute__((visibility("default")))
#    else
        /* We are using this library */
#      define ARCUS_EXPORT __attribute__((visibility("default")))
#    endif
#  endif

#  ifndef ARCUS_NO_EXPORT
#    define ARCUS_NO_EXPORT __attribute__((visibility("hidden")))
#  endif
#endif

#ifndef ARCUS_DEPRECATED
#  define ARCUS_DEPRECATED __attribute__ ((__deprecated__))
#endif

#ifndef ARCUS_DEPRECATED_EXPORT
#  define ARCUS_DEPRECATED_EXPORT ARCUS_EXPORT ARCUS_DEPRECATED
#endif

#ifndef ARCUS_DEPRECATED_NO_EXPORT
#  define ARCUS_DEPRECATED_NO_EXPORT ARCUS_NO_EXPORT ARCUS_DEPRECATED
#endif

#if 0 /* DEFINE_NO_DEPRECATED */
#  ifndef ARCUS_NO_DEPRECATED
#    define ARCUS_NO_DEPRECATED
#  endif
#endif

#endif /* ARCUS_EXPORT_H */
```
Everything is commented out. It then fails to link Arcus into the examples, because it can't find a symbol even though that symbol is marked with ARCUS_EXPORT.

**Comment by onitake** [comment]
I'm a bit confused. What do you mean with "everything is commented out"?
Unless you define `ARCUS_STATIC_DEFINE`, `ARCUS_EXPORT` will be set to `__attribute__((visibility("default")))`, while `ARCUS_NO_EXPORT`will be `__attribute__((visibility("hidden")))`.

I'm not sure if I tested the examples though, hold on...

**Comment by Ghostkeeper** [comment]
> What do you mean with "everything is commented out"?

There is a `#` character before every line.

**Comment by onitake** [comment]
You realize this is C++ code and these are C preprocessor directives, right? :slightly_smiling_face:

**Comment by onitake** [comment]
In case you're wondering about the spaces between the # and the directives, they actually don't matter.
Whitespace before and after the # is ignored by the preprocessor. [1]

The fact that you're getting errors is still an issue, but most likely not caused by the generated header file. Which compiler/OS/CPU are you testing with?

I reapplied the patch against HEAD and it works fine on my machine. Here's a PR to try for yourself: https://github.com/Ultimaker/libArcus/pull/98

[1] https://gcc.gnu.org/onlinedocs/cpp/The-preprocessing-language.html#The-preprocessing-language

**Comment by Ghostkeeper** [comment]
> You realize this is C++ code and these are C preprocessor directives, right?

:laughing: Yeah I guess the spaces between the `#` and the keywords threw me off. Sorry!

I'm trying on GCC 8.3.0, Ubuntu 19.04, Intel Core i7-5650U CPU (2.20GHz). I'll try your PR, thank you!

**Comment by Ghostkeeper** [comment]
This is now implemented through https://github.com/Ultimaker/libArcus/pull/98. Thanks!
